plugins/autoid/plugin.js

```diff
@@ -158,6 +160,209 @@
+          if ( !uploadInitiallyHidden )
+            dialog.hidePage( 'upload' );
+        }
+
+        for ( var i = 0; i < partIds.length; i++ ) {
+          var element = dialog.getContentElement( 'info', partIds[ i ] );
+          if ( !element )
+            continue;
+
+          element = element.getElement().getParent().getParent();
+          if ( partIds[ i ] == typeValue + 'Options' )
+            element.show();
+          else
+            element.hide();
+        }
+
+        dialog.layout();
+      }
+
+      function createHeadingLinkContent() {
+        var headings;
+
+        return {
+          type: 'vbox',
+          id: 'headingOptions',
+          children: [ {
+            type: 'select',
+            id: 'heading',
+            label: 'Select a Heading',
+            style: 'width : 100%;',
+            items: [ [ '' ] ],
+            setup: function() {
+              this.clear();
+              this.add( [ '' ] );
+
+              headings = findAllHeadings();
+
+              if (headings && headings.count()) {
+                this.getElement().show();
+
+                for (var i = 0; i < headings.count(); i++) {
+                  var heading = headings.getItem(i);
+                  this.add( heading.getText(), heading.getId() );
+                }
+              }
+              else {
+                this.getElement().hide();
+              }
+            },
+            commit: function(data) {
+              if (!data.heading)
+                data.heading = {};
+
+              data.heading.id = this.getValue();
+              // set the data type to 'anchor' since the procedure is the same
+              // for headings.  This way, we don't have to modify the main
+              // 'plugin.js' for 'link' to handle a new link type.
+              if (data.heading.id)
+                data.type = 'anchor';
+            }
+          },
+          {
+            type: 'html',
+            id: 'noHeadings',
+            style: 'text-align: center;',
+            html: '<div role="note" tabIndex="-1">' + CKEDITOR.tools.htmlEncode( '(No headings available in the document)' ) + '</div>',
+            // Focus the first element defined in above html.
+            focus: true,
+            setup: function() {
+              this.getElement()[ headings && headings.count() ? 'hide' : 'show' ]();
+            }
+          } ]
+        }
+      }
+
+      function modifiedOnOk() {
+        var data = {},
+          plugin = CKEDITOR.plugins.link;
+
+        // Collect data from fields.
+        this.commitContent( data );
+
+        // if there is a heading id and not an anchor id, the user is creating
+        // a link to a heading.  We set value of the heading id to the anchor.id
+        // property so it will create the new link without having to
+        // modify the plugin further
+        if (data.heading.id && !data.anchor.id)
+          data.anchor.id = data.heading.id;
```

Comment:
This is a bit challenging.  I see this part is different.  Is anything else different?  

It's too bad there's not a better way to hook into the "after commit"